pilot/pkg/serviceregistry/serviceentry/servicediscovery.go

```diff
@@ -398,72 +392,81 @@ func (s *ServiceEntryStore) WorkloadInstanceHandler(wi *model.WorkloadInstance,
-			delete(s.workloadInstancesByName, k)
+			s.workloadInstances.delete(k)
 		}
 	default: // add or update
-		if old, exists := s.workloadInstancesByName[k]; exists {
+		if old := s.workloadInstances.get(k); old != nil {
 			if old.Endpoint.Address != wi.Endpoint.Address {
 				addressToDelete = old.Endpoint.Address
 			}
-
 			// If multiple k8s services select the same pod or a service has multiple ports,
 			// we may be getting multiple events ignore them as we only care about the Endpoint IP itself.
 			if model.WorkloadInstancesEqual(old, wi) {
 				// ignore the update as nothing has changed
 				redundantEventForPod = true
 			}
 		}
-		s.workloadInstancesByName[k] = wi
+		s.workloadInstances.update(wi)
 	}
-	// We will only select entries in the same namespace
-	entries := s.seWithSelectorByNamespace[wi.Namespace]
-	s.storeMutex.Unlock()
 
-	// nothing useful to do.
-	if len(entries) == 0 || redundantEventForPod {
+	if redundantEventForPod {
+		s.mutex.Unlock()
+		return
+	}
+
+	// We will only select entries in the same namespace
+	cfgs, _ := s.store.List(gvk.ServiceEntry, wi.Namespace)
+	if len(cfgs) == 0 {
+		s.mutex.Unlock()
 		return
 	}
 
 	log.Debugf("Handle event %s for service instance (from %s) in namespace %s", event,
 		wi.Endpoint.Address, wi.Namespace)
 	instances := []*model.ServiceInstance{}
 	instancesDeleted := []*model.ServiceInstance{}
-	for _, se := range entries {
+	for _, cfg := range cfgs {
+		se := cfg.Spec.(*networking.ServiceEntry)
 		workloadLabels := labels.Collection{wi.Endpoint.Labels}
-		if !workloadLabels.IsSupersetOf(se.entry.WorkloadSelector.Labels) {
+		if se.WorkloadSelector == nil || !workloadLabels.IsSupersetOf(se.WorkloadSelector.Labels) {
 			// Not a match, skip this one
 			continue
 		}
-		instance := convertWorkloadInstanceToServiceInstance(wi.Endpoint, se.services, se.entry)
+		seNamespacedName := types.NamespacedName{Namespace: cfg.Namespace, Name: cfg.Name}
+		services := s.services.getServices(seNamespacedName)
+		instance := convertWorkloadInstanceToServiceInstance(wi.Endpoint, services, se)
 		instances = append(instances, instance...)
 		if addressToDelete != "" {
 			for _, i := range instance {
 				di := i.DeepCopy()
 				di.Endpoint.Address = addressToDelete
 				instancesDeleted = append(instancesDeleted, di)
 			}
+			s.serviceInstances.deleteServiceEntryInstances(seNamespacedName, key)
+		} else {
+			s.serviceInstances.updateServiceEntryInstancesPerConfig(seNamespacedName, key, instance)
 		}
 	}
-
 	if len(instancesDeleted) > 0 {
-		s.deleteExistingInstances(key, instancesDeleted)
+		s.serviceInstances.deleteInstances(key, instancesDeleted)
 	}
 
-	if event != model.EventDelete {
-		s.updateExistingInstances(key, instances)
+	if event == model.EventDelete {
+		s.serviceInstances.deleteInstances(key, instances)
 	} else {
-		s.deleteExistingInstances(key, instances)
+		s.serviceInstances.updateInstances(key, instances)
 	}
+	s.mutex.Unlock()
 
 	s.edsUpdate(instances, true)
```

Comment:
Slight concern we need edsUpdate inside the lock. I will look at other usages

Reply:
I think not. In edsUpdate we extract the keys. Then in edsUpdateByKeys, we fetch the instances in a lock and fetch the latest instances by key. However, we then release the lock and call EDSUpdate.

One concern might be multiple threads call edsUpdateByKeys:

1. Thread one fetches allInstances@v1
2. Thread two fetches allInstances@v2
3. Thread 2 calls EDSUpdate@v2
4. Thread 1 calls EDSUpdate@v1
Then we lose the v2 update.

WDYT?

This is an existing problem in the current code though

Reply:
Yeah looks like a race. 
I guess previously it is the performance consideration that it is not protected.

Reply:
Need to think about it carefully, maybe we can run one thread updating eds.

Reply:
done